ndb/src/google/cloud/ndb/eventloop.py

```diff
@@ -30,16 +31,228 @@
+
+    def queue_rpc(self, rpc, callback=None, *args, **kwds):
+        """Schedule an RPC with an optional callback.
+
+        The caller must have previously sent the call to the service.
+        The optional callback is called with the remaining arguments.
+
+        NOTE: If the rpc is a MultiRpc, the callback will be called once
+        for each sub-RPC.  TODO: Is this a good idea?
+        """
+        # TODO Integrate with gRPC
         raise NotImplementedError
 
+    def add_idle(self, callback, *args, **kwargs):
+        """Add an idle callback.
+
+        An idle callback can return True, False or None.  These mean:
+
+        - None: remove the callback (don't reschedule)
+        - False: the callback did no work; reschedule later
+        - True: the callback did some work; reschedule soon
+
+        If the callback raises an exception, the traceback is logged and
+        the callback is removed.
+        """
+        self.idlers.append((callback, args, kwargs))
+
+    def run_idle(self):
+        """Run one of the idle callbacks.
+
+        Returns:
+          True if one was called, False if no idle callback was called.
+        """
+        if not self.idlers or self.inactive >= len(self.idlers):
+            return False
+        idler = self.idlers.popleft()
+        callback, args, kwargs = idler
+        _logging_debug('idler: %s', callback.__name__)
+        result = callback(*args, **kwargs)
+
+        # See add_idle() for meaning of callback return value.
+        if result is None:
+            _logging_debug('idler %s removed', callback.__name__)
+        else:
+            if result:
+                self.inactive = 0
+            else:
+                self.inactive += 1
+            self.idlers.append(idler)
+        return True
+
+    def _run_current(self):
+        """Run one current item.
+
+        Returns:
+            True if one was called, False if no callback was called.
+        """
+        if not self.current:
+            return False
+
+        self.inactive = 0
+        callback, args, kwargs = self.current.popleft()
+        _logging_debug('nowevent: %s', callback.__name__)
+        callback(*args, **kwargs)
+        return True
+
+    def run0(self):
+        """Run one item (a callback or an RPC wait_any).
+
+        Returns:
+          A time to sleep if something happened (may be 0);
+          None if all queues are empty.
+        """
+        if self._run_current() or self.run_idle():
+            return 0
+
+        delay = None
+        if self.queue:
+            delay = self.queue[0][0] - self.clock.now()
+            if delay <= 0:
+                self.inactive = 0
+                _, callback, args, kwargs = self.queue.pop(0)
+                _logging_debug('event: %s', callback.__name__)
+                callback(*args, **kwargs)
+                # TODO: What if it raises an exception?
+                return 0
+
+        if self.rpcs:
+            raise NotImplementedError
```

Comment:
I have been adding a note about these "partially" done methods in #6251.

Reply:
Done.